tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+            let mut propagation_simulators = Vec::with_capacity(num_blocks);
+
+            // First block does not get a carry from
+            propagation_simulators.push(self.key.create_trivial(0));
+            for block in propagation_cum_sums.drain(..) {
+                if propagation_simulators.len() % grouping_size == 0 {
+                    groupings_pgns.push(block);
+                    // The first block in each grouping has its simulator set to 0
+                    // because it always receives any input borrow that may be generated from
+                    // previous grouping
+                    propagation_simulators.push(self.key.create_trivial(1));
+                } else {
+                    propagation_simulators.push(block);
+                }
+            }
+
+            let mut prepared_blocks = shifted_blocks;
+            prepared_blocks
+                .iter_mut()
+                .zip(propagation_simulators.iter())
+                .for_each(|(block, simulator)| {
+                    crate::core_crypto::algorithms::lwe_ciphertext_add_assign(
+                        &mut block.ct,
+                        &simulator.ct,
+                    );
+                });
+
+            match requested_flag {
+                OutputFlag::None => {}
+                OutputFlag::Overflow => {
+                    let block = output_flag.as_mut().unwrap();
+                    self.key
+                        .unchecked_add_assign(block, &propagation_simulators[num_blocks - 1]);
+                }
+                OutputFlag::Carry => {
+                    let block = output_flag.as_mut().unwrap();
+                    self.key
+                        .unchecked_add_assign(block, &propagation_simulators[num_blocks - 1]);
+                }
+            }
+
+            (prepared_blocks, groupings_pgns)
+        };
+
+        // Third step: resolving carry propagation between the groups
+        let resolved_carries = if groupings_pgns.is_empty() {
+            vec![self.key.create_trivial(0)]
+        } else if use_sequential_algorithm_to_resolved_grouping_carries {
+            let luts = (0..grouping_size - 1)
+                .map(|index| {
+                    self.key.generate_lookup_table(|propa_cum_sum_block| {
+                        let carry = propa_cum_sum_block & (1 << (index + 1));
+                        u64::from(carry != 0)
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            groupings_pgns.rotate_left(1);
+            let mut resolved_carries =
+                vec![self.key.create_trivial(0), groupings_pgns.pop().unwrap()];
+            for chunk in groupings_pgns.chunks(grouping_size - 1) {
+                let mut cum_sums = chunk.to_vec();
+                self.key
+                    .unchecked_add_assign(&mut cum_sums[0], resolved_carries.last().unwrap());
+
+                for i in [1, 2] {
```

Comment:
I don't think I understand what this loop does

Reply:
Its a cumulative sum written in the weirdest way possible because .windows_mut does not exists

But i've rewritten it so that it should be simpler